sshfs/file.py

```diff
@@ -81,6 +81,9 @@ async def _open_file(self):
     def readable(self):
         return "r" in self.mode
 
+    def seekable(self):
+        return "r" in self.mode or "w" in self.mode
```

Comment:
Just to double check: can it be also '+' (update)? 

Also, is it actually enforced? How do we actually open files, write to them? E.g do we always seek to the end in the append mode on each write?

Reply:
> Just to double check: can it be also '+' (update)?
> 
> Also, is it actually enforced? How do we actually open files, write to them? E.g do we always seek to the end in the append mode on each write?

Note that you have this code on line 21:

```python3
        # TODO: support r+ / w+ / a+
        if mode not in {"rb", "wb", "ab"}:
```

Which makes the questions for `+` currently moot and untestable, but I'd also like to see support for +-mode. I'd also like to see support for open via [flags](https://github.com/ronf/asyncssh/blob/46636d6f18b221141e659e2562c49b4a271befdb/asyncssh/sftp.py#L185) like in asnycssh, but this is not that important. Coming back to your question, yes + mode should also be seekable. I mean, in the end, it only means that it can read and write instead of only doing one of both.

Append mode is more difficult for me to decide on because I have never used it. It seems to be enforced inside [`asyncssh.sftp.SFTPFile`](https://github.com/ronf/asyncssh/blob/46636d6f18b221141e659e2562c49b4a271befdb/asyncssh/sftp.py#L3114) and it seems to actually allow for seeking in general, but it only has an effect for read. I am not sure where it is enforced exactly, but there is [this](https://github.com/ronf/asyncssh/blob/46636d6f18b221141e659e2562c49b4a271befdb/asyncssh/sftp.py#L3310) comment in the `write` method.

Some code for testing assumptions:

```python3
import os
import sshfs

path = "foo"

if os.path.exists(path):
    os.remove(path)
with open(path, "wb") as file:
    file.write(b"123456")
with open(path, "ab") as file:
    file.write(b"789")
    assert file.tell() == 9
    file.seek(3)
    assert file.tell() == 3
    # file.read(1)  # io.UnsupportedOperation: read
    file.write(b"xyz")
    # Note that tell does NOT return 12 here as one might be inclined to think!
    # Instead of doing implicit seeks before each write, append simply always appends.
    # https://docs.python.org/release/3.7.2/library/functions.html#open
    # > 'a' for appending (which on some Unix systems, means that all writes
    # > append to the end of the file regardless of the current seek position).
    assert file.tell() == 6
with open(path, "ab+") as file:
    file.seek(6)
    assert file.tell() == 6
    assert file.read() == b"789xyz"
with open(path, "rb") as file:
    assert file.read() == b"123456789xyz"

fs = sshfs.SSHFileSystem("127.0.0.1", port=2022)

if fs.exists(path):
    fs.delete(path)
with fs.open(path, "wb") as file:
    file.write(b"123456")
with fs.open(path, "ab") as file:
    file.write(b"789")
    assert file.tell() == 9
    file.seek(3)
    assert file.tell() == 3
    file.write(b"xyz")
    # This is different from the built-in open file.
    assert file.tell() == 12
with fs.open(path, "ab+") as file:
    file.seek(6)
    assert file.tell() == 6
    assert file.read() == b"789xyz"
with fs.open(path, "rb") as file:
    assert file.read() == b"123456789xyz"
```

For the test with `a+` via sshfs, I simply commented out the mode check mentioned above.

Note that I would have delegated to `asyncssh.sftp.SFTPFile.seekable` if there were such a method in the first place ...

I think it might be more correct to return true for seekable in the case of `a+` and all other modes except for simply `a`, for which I am conflicted. `seek` theoretically works in the way that it does not raise an exception and calling tell after it will return the seeked-to offset, but read is not allowed in pure `a` mode and writing will always be done to the end...

Reply:
But as a+ is not yet supported, the current commit should be sufficiently correct in my opinion.

Reply:
I have added asserts for `seekable` to the Python file object and it returns true even for `a` mode. Therefore, I have pushed the change to simply always return True from inside seekable.